File: tutorials/dynamic-receiving-with-mpi-probe-and-mpi-status/code/probe.c
// Author: Wes Kendall
// Copyright 2011 www.mpitutorial.com
// This code is provided freely with the tutorials on mpitutorial.com. Feel
// free to modify it for your own use. Any distribution of the code must
// either provide a link to www.mpitutorial.com or keep this header intact.
//
// Example of using MPI_Probe to dynamically allocated received messages
//
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int main(int argc, char** argv) {
  MPI_Init(NULL, NULL);

  int world_size;
  MPI_Comm_size(MPI_COMM_WORLD, &world_size);
  if (world_size != 2) {
    fprintf(stderr, "Must use two processes for this example\n");
    MPI_Abort(MPI_COMM_WORLD, 1);
  }
  int world_rank;
  MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);

  int number_amount;
  if (world_rank == 0) {
    const int MAX_NUMBERS = 100;
    int numbers[MAX_NUMBERS];
    // Pick a random amont of integers to send to process one
    srand(time(NULL));
    number_amount = (rand() / (float)RAND_MAX) * MAX_NUMBERS;
    // Send the amount of integers to process one
    MPI_Send(numbers, number_amount, MPI_INT, 1, 0, MPI_COMM_WORLD);
    printf("0 sent %d numbers to 1\n", number_amount);
  } else if (world_rank == 1) {
    MPI_Status status;
    // Probe for an incoming message from process zero
    MPI_Probe(0, 0, MPI_COMM_WORLD, &status);
    // When probe returns, the status object has the size and other
    // attributes of the incoming message. Get the size of the message.
    MPI_Get_count(&status, MPI_INT, &number_amount);
    // Allocate a buffer just big enough to hold the incoming numbers
    int* number_buf = (int*)malloc(sizeof(int) * number_amount);
    // Now receive the message with the allocated buffer
    MPI_Recv(number_buf, number_amount, MPI_INT, 0, 0, MPI_COMM_WORLD,
             MPI_STATUS_IGNORE);
    printf("1 dynamically received %d numbers from 0.\n",
           number_amount);
    free(number_buf);
  }
  MPI_Finalize();
}

File: tutorials/mpi-fortran/code/probe.f90
program probe
  use mpi_f08

  implicit none

  integer              :: world_rank, world_size
  integer, parameter   :: MAX_NUMBERS=100
  integer, allocatable :: numbers(:)
  integer              :: number_amount
  type(MPI_Status)     :: recv_status

  real :: r

  call MPI_INIT()
  call MPI_COMM_SIZE(MPI_COMM_WORLD, world_size)
  call MPI_COMM_RANK(MPI_COMM_WORLD, world_rank)


  if (world_rank .eq. 0) then
    ! Pick a random amount of integers to send to process one
    call random_seed()

    call random_number(r)
    number_amount = int(r * real(MAX_NUMBERS))
    allocate(numbers(MAX_NUMBERS))
    ! Send the random amount of integers to process one
    call MPI_SEND(numbers, number_amount, MPI_INT, 1, 0, &
                  MPI_COMM_WORLD)
    print '("0 sent ", I0, " numbers to 1")', number_amount
  else if (world_rank .eq. 1) then
    ! Probe for an incoming message from process zero
    call MPI_PROBE(0, 0, MPI_COMM_WORLD, recv_status)

    ! When probe returns, the status object has the size and other
    ! attributes of the incoming message. Get the message size
    call MPI_Get_count(recv_status, MPI_INT, number_amount)

    ! Allocate a buffer to hold the incoming numbers
    allocate(numbers(number_amount))

    ! Now receive the message with the allocated buffer
    call MPI_RECV(numbers, number_amount, MPI_INT, 0, 0, &
                  MPI_COMM_WORLD, MPI_STATUS_IGNORE)
    print '("1 dynamically received ", I0, " numbers from 0.")', &
      number_amount
    deallocate(numbers)
  end if

  call MPI_FINALIZE()

end program
